README.md
# Installing Starwatch 
You can add Starwatch to an existing project pretty easily with these steps.

* Prepare your Mongo DB. 
   Set up a MongoDB of your choice, or sign up for a free hosted on http://www.mongolab.com. Note its connection info, and create a new collection called 'log'.
   

* Add the Starwatch-Resources files to your project.

* A few of them are not yet ARC-compliant. Add the -fno-objc-arc compilation flag to your Target's Build Phases under Compile Sources to the following files:
    - TwentyNineDBHelper.m
    - FMDatabase.m
    - FMResultSet.m
    - JSONKit.m
    
* Update the Starwatch-Resources/SWConstants.M file's mongo_settings dictionary values with the connectivity info you got in the first step.

* Add the following frameworks and libraries to your target's Build Phases -> Link Binary with Libraries'.  All are required.
      System Configuration framework
      CFNetwork framework
      libsqlite3.0.dylib
      
      
* The mongo-c-driver library requires this additional compiler flag to be set.  Under your target's Build Settings, click the Add Build Setting at the bottom.  The key for this setting will be "_MONGO_USE_GETADDRINFO" and the value will be 1.

 
* Do yourself a favor early on and turn on actions_to_stderr in your SWConstant.m file.  This will echo every action tracked by Starwatch to your stderr log for easy testing.  Otherwise, it happens in the background and you'd have to inspect the sqlite database to confirm record collection, which is really tough when you're working directly on a device.
      
Now your code is ready to begin using Starwatch.
      
      
# Starwatch: Basic
The easiest way to get started with Starwatch is to log the most basic actions.  The "Starwatch-Basic" project sends one hunk of information, called the "INFO" action. It will dump the time, what version of iOS the user is using, what type of device they have, their device timezone as a string, and the number of times they've opened the app.  

To use just the INFO packet in your project, include "SWCUtility.h" and insert the following snippets into your AppDelegate file.

in didFinishLaunchingWithOptions:
      
      /*** Starwatch BASIC ***/
      // This will make sure our DB's are in the correct place,
      // we've begun tracking actions against this unique device id,
      // and increments the number of times this user has opened the app.
      [SWCUtility begin];


      // Log the "INFO" action.
      // This method takes in a dictionary of your custom key-value pairs
      // and, along with general information about this device and user,
      // prepares to send it to the remote db.
      [SWCUtility logInfo:[NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:
                                                               [NSString stringWithFormat:@"%d", [SWCUtility getNumOpens]],
                                                               nil]
                                                      forKeys:[NSArray arrayWithObjects:
                                                               @"num_opens",
                                                               nil]
                           ]
       ];
   
   
You also need to add something into the Enter Background method so that the data is actually sent.  I've added this into applicationDidEnterBackground:
      /*** Starwatch BASIC ***/
      [SWCUtility send_data];
   
   
   
To test, confirm your actions_to_stderr in SWConstants is set to YES, and when you open or close the app in the simulator or on a device, you should see info actions logged.  Example:

      2012-10-13 17:24:04.406 Starwatch-Plain[7959:11303] 
      [SWCUtility] {"name":"","action":"info","metadata":"{\"app_version\":\"1.0\",\"device\":\"iPad Simulator\",\"device_version\":\"iPad Simulator\",\"timezone\":\"America/New_York\",\"ios\":\"6.0\",\"num_opens\":\"2\"}","global_id":""}
   
When you minimize the app in the simulator or device (NOT by Stopping the running app in Xcode), it will trigger the send-data action.  Turn on sending of data by editing the send_stats value in SWConstants and look at your Mongo database to ensure you're collecting data.  Read "Parsing Collected Data" to work with the data you've collected.


   
   
   
# Starwatch: Advanced

   * ADD THE METHODS TO YOUR CODE.

      * In the App Delegate, add to the top:

      `#import "StarWatchDatastoreUtility.h"
      #import "SWCUtility.h"`

      * In didFinishLaunchingWithOptions, add:

         `// Fire up the stats engine and log the action of booting up.
         [StarWatchDatastoreUtility setupDatabaseAndIdentifier];
         [SWCUtility logAppStart];`

      * in applicationWillResignActive, add:
         `[SWCUtility logAppActive];`

      * in applicationDidEnterBackground, add:
         `[SWCUtility logAppEnd];`
      
then, use the SW* controllers instead of the core controllers.



# Parsing Collected Data and Presenting the Results
The `web_app/` folder is a python application that runs the tornado webserver on your local machine.  Installing tornado and pymongo is straightforward (`easy_install tornado`; `easy_install pymongo`), but if you haven't run this stack locally through virtualenv's and such, I recommend running https://github.com/nataliepo/HelloTornado just to test that your environment is ready.

## 
## Initializing
First, copy the `web_app/settings.py.default` file to settings.py.  Then, edit the `settings.py` file to reflect your mongodb settings, particularly in the MONGO_SETTINGS section.  The script should create the Collection definitions (session, view, device, feedback, etc) if they don't already exist.



## Info / "Basic" parsing
The `parse_logs.py` python script will aggregate unique device information collected and place it in your devices collection on mongo so that the webapp can present it for you.

To parse just the info logs, which is what was used in the Starwatch-Basic project, run this:
      python parse_log.py info
      
This sets a flag in each log record that the log has been seen so it won't be re-evaluated.  If you ever want to clear the info logs you've seen and re-parse the data, 
      python parse_log.py info-purge
will safely mark all logs of action INFO as not being seen so you can run parse_log.py info on it again to recollect the data.


## Advanced Parsing
We'll be releasing more parsing methods soon! (We need to test this a bit more, then document how it works, since it's very specialized to our platform.)



# License
Copyright © 2012, 29th Street Publishing, and Distributed under
the Artistic License.


CHANGES.md
Starwatch  - Changelog
==========================

Unreleased
-------------
* Included basic collection (SWCUIViewController, SWCUIButton) and sending (SWSender) mechanisms
* Ensured collected data was sent to a mongolab db via mongo-c-driver
* Counted number of app opens as a part of the app's "info" log action



1.0 (2012-10-17)
----------------
Initial release of the library


iOS/lib/SWCUtility.m
//
//  SWCUtility.m
//  StarWatch
//
//  Created by Natalie Podrazik on 1/17/12.
//  Copyright (c) 2012 29th Street Publishing. All rights reserved.
//

#import "SWCUtility.h"
#import "SWCViewController.h"
#import "SWSender.h"

#import "JSONKit.h"

#import "SWConstants.h"

#import "TwentyNineDBHelper.h"


#define STARWATCH_KEY_NUM_OPENS @"STARWATCH-NUM_OPENS"

#define STATS_SETTINGS_TABLE @"settings"
#define SETTINGS_KEY_DEVICE_ID @"device"
#define STARWATCH_LOG_TABLE @"log"

#define IPHONE_4_HEIGHT 480
#define RETINAL_RESOLUTION 2.0


@implementation SWCUtility

+(void)logError:(NSString*)name
         action:(NSString*)action
      global_id:(NSString*)global_id
       metadata:(NSString*)metadata {
    // for now, this is the same as logging an action.
    [self logAction:name
             action:action
          global_id:global_id
           metadata:metadata];
}

+(void)logAction:(NSString*)name
          action:(NSString*)action
       global_id:(NSString*)global_id
        metadata:(NSString*)metadata {
    
    [SWCUtility logDictionary:[SWCUtility buildLogDictionary:name global_id:global_id action:action metadata:metadata]];
    
}

// we're going to get all this information from the device.
// note: all the information is in the metadata, which will be parsed out later.
+(void)logInfo:(NSDictionary*)customInfo {
    
    NSMutableDictionary * info = [[NSMutableDictionary alloc] initWithDictionary:customInfo];
    
    [info setObject:[[UIDevice currentDevice] localizedModel]
             forKey:@"device"];
    
    [info setObject:[SWCUtility estimateDeviceType]
             forKey:@"device_version"];
    
    
    [info setObject:[[UIDevice currentDevice] systemVersion]
             forKey:@"ios"];
    
    
    [info setObject:[[[NSBundle mainBundle] infoDictionary] objectForKey:(NSString*)kCFBundleVersionKey]
             forKey:@"app_version"];
    
    
    [info setObject:[[NSTimeZone systemTimeZone] name]
             forKey:@"timezone"];
    
    // now this info hunk as metadata.  let the client do more work.
    [SWCUtility logAction:@""
                   action:SW_ACTION_INFO
                global_id:@""
                 metadata:[info JSONString]];

    
    
    
    info = nil;
    
}


+(NSString*)estimateDeviceType {
    NSString * base_model = [[UIDevice currentDevice] localizedModel];
    CGFloat screenScale = [[UIScreen mainScreen] scale];

    
    // iPhone....
    if ([base_model isEqualToString:@"iPhone"]) {
        
        // First hint:
        //  iPhones 4 and below are sized 320x480,
        //  iPhones 5 (and beyond!) are taller.
        CGRect bounds = [[UIScreen mainScreen] bounds];
        
        if (bounds.size.height > IPHONE_4_HEIGHT) {
        
            return [NSString stringWithFormat:@"%@ 5", base_model];
        }
        
        // second hint: iPhones 3 and below are not retinal.
        if (screenScale < 2.0) {
            return [NSString stringWithFormat:@"%@ 3-", base_model];
        }
    
        // otherwise, it's a 4.
        return [NSString stringWithFormat:@"%@ 4", base_model];    
    }


    // otherwise, it's an iPad.
    else if ([base_model isEqualToString:@"iPad"]) {
        // first hint: iPad 1 is the only one that does not have a camera.
        if (![UIImagePickerController isSourceTypeAvailable:UIImagePickerControllerSourceTypeCamera]) {
            return [NSString stringWithFormat:@"%@ 1", base_model];
        }

        // hint number 2: iPad 2's are not retinal.
        if (screenScale < 2.0) {
            return [NSString stringWithFormat:@"%@ 2", base_model];
        }


        // everything else, we assume it's an iPad 3.
        return [NSString stringWithFormat:@"%@ 3", base_model];
    }
    else if ([base_model isEqualToString:@"iPod touch"]) {
        return base_model;
    }


    return [NSString stringWithFormat:@"%@", base_model];

}


+(NSDictionary*)buildLogDictionary:(NSString*)name 
                         global_id:(NSString*)global_id
                            action:(NSString*)action  
                          metadata:(NSString*)metadata {
    
    // quick parameter check.
    // can't make this an array with objects because one might be nil,
    // defeating the purpose of that efficiency; boo.
    if (name == nil) {
        name = @"";
    }
    
    if (global_id == nil) {
        global_id = @"";
    }
    
    if (action == nil) {
        action = @"";
    }
    
    if (metadata == nil) {
        metadata = @"";
    }
    

    return [NSDictionary dictionaryWithObjects:[NSArray arrayWithObjects:name, 
                                                action,
                                                global_id,
                                                metadata, 
                                                nil]
                                       forKeys:[NSArray arrayWithObjects:@"name", 
                                                @"action", 
                                                @"global_id",
                                                @"metadata",
                                                nil]
            ];
}


/* 
 these actions are special - we always want to look for the same start-up
 and ending strings.
 */
+(void)logAppStart {
    // you can't report orientation data until you load a view.
    
    [SWCUtility logDictionary:[SWCUtility buildLogDictionary:SW_VIEW_APP_DELEGATE
                                                   global_id:@""
                                                      action:SW_ACTION_START_APP 
                                                    metadata:@""]];
    
    [SWCUtility incrementNumOpens];
    
    

}

+(void)logAppEnd {
    [SWCUtility logDictionary:[SWCUtility buildLogDictionary:SW_VIEW_APP_DELEGATE
                                                   global_id:@""
                                                      action:SW_ACTION_ENTERED_BACKGROUND
                                                    metadata:@""]];
    
    [SWSender sendData];
    
}

+(void)logAppCrash {
    [SWCUtility logDictionary:[SWCUtility buildLogDictionary:SW_VIEW_APP_DELEGATE
                                                   global_id:@""
                                                      action:SW_ACTION_CRASH
                                                    metadata:@""]];
}



+(void)logAppActive {
    [SWCUtility logDictionary:[SWCUtility buildLogDictionary:SW_VIEW_APP_DELEGATE
                                                   global_id:@""
                                                      action:SW_ACTION_BECAME_ACTIVE
                                                    metadata:@""]];
}

+(int)getNumOpens {    
    if ([[NSUserDefaults standardUserDefaults] stringForKey:STARWATCH_KEY_NUM_OPENS] == nil) {
        [[NSUserDefaults standardUserDefaults] setValue:[NSString stringWithFormat:@"%d", 0]
                                                 forKey:STARWATCH_KEY_NUM_OPENS];
        
        [[NSUserDefaults standardUserDefaults] synchronize];

    }
    
    return [[[NSUserDefaults standardUserDefaults] objectForKey:STARWATCH_KEY_NUM_OPENS] intValue];
}

+(void)incrementNumOpens {
    
    int num_opens = [SWCUtility getNumOpens];
    num_opens += 1;
    
    [[NSUserDefaults standardUserDefaults] setValue:[NSString stringWithFormat:@"%d", num_opens]
                                             forKey:STARWATCH_KEY_NUM_OPENS];
}

+(NSString*)databaseName {         
    return [SWConstants database_name];
}


+(void)logDictionary:(NSDictionary*)meta_dictionary {
    
    // write the record translated to json.
           [SWCUtility logString:[meta_dictionary JSONString]];
   
    
    
    /* don't do this asynchronously.
     * 
     dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, (unsigned long)NULL), ^(void) {
     */
    
    //[SWSender sendData];
    
    //});
}



+(void)logString:(NSString*)metadata {
        

    
    if (![SWConstants collect_stats]) {
        return;
    }
    
    if ([SWCUtility actions_to_stderr]) {
        NSLog(@"\n[SWCUtility] %@\n", metadata);
    }
    
   
    NSDateFormatter * formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat:@"yyyyMMdd HH:mm:ss"];
    [formatter setTimeZone:[NSTimeZone timeZoneWithAbbreviation:@"EST"]];
    NSDate * now = [NSDate dateWithTimeIntervalSinceNow:0];
    
    
    // wrap both parameters in single quotes.
    NSString * readableTimestamp = [TwentyNineDBHelper wrapInSingleQuotes:[formatter stringFromDate:now]];
    metadata = [TwentyNineDBHelper wrapInSingleQuotes:metadata];

    NSMutableArray * cols = [NSMutableArray arrayWithObjects:@"last_updated",
                             @"metadata",
                             nil];
    NSMutableArray * vals = [[NSMutableArray alloc] initWithArray:[NSArray arrayWithObjects:readableTimestamp, metadata, nil]];
    
    
    
    
        
    [TwentyNineDBHelper insertQuery:[SWCUtility databaseName]
                                                 columns:cols 
                                                  values:vals 
                                                   table:@"log"];
        
        
        
    vals = nil;
    cols = nil;
    
    formatter = nil;

    return;
}



// helpful hint: http://stackoverflow.com/questions/2633801/generate-a-random-alphanumeric-string-in-cocoa
+(NSString*)createCustomDeviceIDString {
    
    NSString *letters = @"ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    int length = 24;
    
    srand(time(0));
    
    
    
    NSMutableString *randomString = [NSMutableString stringWithCapacity:length];
    
    for (int i = 0; i < length; i++) {
        [randomString appendFormat: @"%c", [letters characterAtIndex: rand()%[letters length]]];
    }
    
    NSLog(@"\n Device string = %@\n", randomString);
    
    return randomString;
}


+(NSString*)getUniqueDeviceID {
                  
    // attempting:
    //  select key from settings where key='device';
    
         
    NSMutableArray * device_results = [TwentyNineDBHelper queryForColsWithClause:[SWCUtility databaseName]
                                                                   columnsWanted:[NSArray arrayWithObject:@"value"]
                                                                           table:STATS_SETTINGS_TABLE
                                                                     whereClause:[NSString stringWithFormat:@"key='%@'", SETTINGS_KEY_DEVICE_ID]
                                                                    sortByClause:@"id desc"];
    
    if (device_results == nil) {
        return nil;
    }
    
    NSString * value = [[device_results objectAtIndex:0] objectForKey:@"value"];
    
    device_results = nil;
    
    // otherwise, return the first one.
    return value;

}



+(void)confirmCustomDeviceIDExists {
        
    NSString * query = [SWCUtility getUniqueDeviceID];
    
    // if there are no results, create a new record.
    if (query == nil) {        
        
        NSMutableArray * cols = [NSMutableArray arrayWithObjects:@"key", @"value", nil];
        
        NSMutableArray * vals = [[NSMutableArray alloc] initWithArray:[NSArray arrayWithObjects:[TwentyNineDBHelper wrapInSingleQuotes:SETTINGS_KEY_DEVICE_ID],
                                                                       [TwentyNineDBHelper wrapInSingleQuotes:[SWCUtility createCustomDeviceIDString]], nil]];

        
        [TwentyNineDBHelper insertQuery:[SWCUtility databaseName]
                                             columns:cols 
                                              values:vals 
                                               table:STATS_SETTINGS_TABLE];
        
        vals = nil;
        cols = nil;
    }
    
    
    // otherwise, it exists; we're good.
    
    return;    
}


// will return an array of strings, to be repackaged within greater json chunks later.
+(NSMutableArray*)getOlderLogRecords:(int)num_records {
     
    int current_record_count = [SWCUtility num_records];
     
    // in case we're working with just a few records, 
    // readjust the return window.
    if (current_record_count < num_records) {
        num_records = current_record_count;
    }
     
    NSString * device_id = [SWCUtility getUniqueDeviceID];
        
    NSArray * query_results = [TwentyNineDBHelper queryForColsWithClause:[SWCUtility databaseName]
                                                     columnsWanted:[NSArray arrayWithObjects:@"*", nil]
                                                             table:STARWATCH_LOG_TABLE 
                                                       whereClause:nil
                                                      sortByClause:[NSString stringWithFormat:@"id limit %d", num_records]];

        
    if ((query_results != nil) && ([query_results count] > 0)) {
        
        NSMutableArray * packaged_results = [[NSMutableArray alloc] initWithCapacity:[query_results count]];
        for (int i = 0; i < [query_results count]; i++) {
            
            NSDictionary * s = [query_results objectAtIndex:i];
                        
            NSDictionary * metadataObj = [[s objectForKey:@"metadata"] objectFromJSONString];
            
            
            NSArray * objects = [NSArray arrayWithObjects:
                                 [s objectForKey:@"id"], 
                                 [s objectForKey:@"last_updated"],
                                 device_id, 
                                 [s objectForKey:@"metadata"],
                                 [metadataObj objectForKey:@"global_id"],

                                 nil];
            
            NSArray * keys = [NSArray arrayWithObjects:
                              @"id",
                              @"last_updated", 
                              @"device",
                              @"action_data",
                              @"global_id",
                              nil];
            
            [packaged_results addObject:[NSDictionary dictionaryWithObjects:objects                                                 forKeys:keys]];            
        }
        
        query_results = nil;
        
        
        // we should probably free the query results.
        //[query_results release];
        return packaged_results;

    }

         

    return nil;
            
}

+(int)num_records {
    return [TwentyNineDBHelper getCountForTable:[SWCUtility databaseName]
                                          table:@"log" 
                                      condition:nil];
}


+(BOOL) actions_to_stderr {
    return [SWConstants actions_to_stderr];
}

+(void)send_data {
    [SWSender sendData];
}

+(BOOL) send_stats {
    return [SWConstants send_stats];
}

+(void)deleteBatchRecords:(NSMutableArray*)records {
    
    NSMutableString * id_str = [NSMutableString stringWithString:@""];
    
    for (int i = 0; i < [records count] - 1; i++) {
        [id_str appendString:[NSString stringWithFormat:@"%@, ", [records objectAtIndex:i]]];
    }
    
    if ([records count] >= 1) {
        [id_str appendString:[NSString stringWithFormat:@"%@", [records objectAtIndex:([records count] - 1)]]];
    }

    
    [TwentyNineDBHelper deleteRecordsForCondition:[SWCUtility databaseName]
                                            table:STARWATCH_LOG_TABLE 
                                        condition:[NSString stringWithFormat:@"id in (%@)", id_str]];
    
    
    return;
}

+(void)begin {
    // Confirms the little Starwatch DB exists and is in the right place for writing.
    [TwentyNineDBHelper createEditableCopyofDatabaseIfNeeded:[SWCUtility databaseName]
                                              forceOverwrite:YES];
    
    // Also confirm that we've generated a unique id for this device.
    [SWCUtility confirmCustomDeviceIDExists];
    
    // increment the number of opens. starts at zero.
    [SWCUtility incrementNumOpens];

}
@end
